Add LinkabilityChecker tool

JamesNK opened the pull request:
Addresses https://github.com/dotnet/aspnetcore/issues/27384

Replicate https://github.com/dotnet/aspnetcore/tree/5760f2864016081a142642b417db17e5340a1d67/src/Components/WebAssembly/testassets/WasmLinkerTest

I've removed the wasm specific build attributes:

```xml
    <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
    <SelfContained>true</SelfContained>
    <UseMonoRuntime>true</UseMonoRuntime>
    <UsingMicrosoftNETSdkBlazorWebAssembly>true</UsingMicrosoftNETSdkBlazorWebAssembly>
```

For lack of anywhere else to put it, I've placed the new testassets project in Shared. @aspnet-build let me know if you'd like it moved somewhere else.

JamesNK committed:
Add LinkerTest project

JamesNK commented:
@pranavkm After removing the wasm/mono attributes the project generates an error:

```
   "C:\Development\Source\aspnetcore\src\Shared\test\testassets\LinkerTest\LinkerTest.csproj" (default target) (1:7) ->
   (ILLinkTrimProjects target) ->
     ILLink : error IL1009: Assembly reference 'System.Security.Permissions' could not be resolved. [C:\Development\Source\aspnetcore\src\Shared\test\testassets\LinkerTest\LinkerTest.csproj]
     ILLink : error IL1009: Assembly reference 'System.Threading.AccessControl' could not be resolved. [C:\Development\Source\aspnetcore\src\Shared\test\testassets\LinkerTest\LinkerTest.csproj]
```

There are no references to those projects in aspnetcore.

Any idea what is wrong? If not, who should I talk to?

davidfowl commented:
You might want to also get rid of this test https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Hosting/test/FunctionalTests/LinkedApplicationTests.cs#L12

pranavkm committed:
Fixup

pranavkm commented:
@JamesNK I'm gonna guess there's a reference to these assemblies in the graph somewhere (maybe a `TypeForwardedTo`) which requires this assemblies to be available to the trimmer. Adding these package reference appears to help here.

pranavkm commented:
FYI - I force pushed to this branch after rebasing on main

pranavkm committed:
Merge branch 'main' into jamesnk/linkertest

wtgodbe reviewed (approved):
`src/Shared` seems reasonable to me

dougbu commented:
> > For lack of anywhere else to put it, I've placed the new testassets project in Shared. @aspnet-build let me know if you'd like it moved somewhere else.

> `src/Shared` seems reasonable to me

I don't feel strongly about this but what is "shared" about this project❔ Not quite understanding how it fits into this repo in the first place and where it might fit otherwise. I'd feel better if the project were in its own src/ directory than under Shared.

JamesNK commented:
There isn't anything shared. Basically, this is a test project that checks linking warnings for projects. It will end up referencing most shipping projects.

An alternative location I considered was /src/Tools. It is kind of a tool for checking for linker warnings and generating linker suppression.

JamesNK commented:
> FYI - I force pushed to this branch after rebasing on main

You're a ⭐

dougbu commented:
> > > > For lack of anywhere else to put it, I've placed the new testassets project in Shared. @aspnet-build let me know if you'd like it moved somewhere else.

> > > `src/Shared` seems reasonable to me

> > I don't feel strongly about this but what is "shared" about this project❔ Not quite understanding how it fits into this repo in the first place and where it might fit otherwise. I'd feel better if the project were in its own src/ directory than under Shared.

> There isn't anything shared. Basically, this is a project that checks linking warnings in shipping projects. It will end up referencing most shipping projects. It's generic, but it isn't itself shared.

> An alternative location I considered was /src/Tools. It is kind of a tool for checking for linker warnings and generating linker suppression.

src/Tools/ would work for me. Other options might be src/Testing/test/ or src/CheckLinkability/.

JamesNK committed:
Move project

pranavkm commented:
@JamesNK / @dougbu I was playing around with this change and came up with https://github.com/pranavkm/aspnetcore/commit/0a96a3dedad688de83d5790a0911d184df0af0dc. The idea is to use `ProjectReferenceProvider` items to find projects that are part of the shared framework and are trimmable. This negates the need to maintain a separate list of trimmable projects.

Thoughts?

dougbu commented:
> I &hellip; came up with https://github.com/pranavkm/aspnetcore/commit/0a96a3dedad688de83d5790a0911d184df0af0dc. The idea is to use ProjectReferenceProvider items to find projects that are part of the shared framework and are trimmable. This negates the need to maintain a separate list of trimmable projects.

The idea sounds reasonable though I suggest things may be simpler if the metadata were collected when we run CodeGen.proj rather than added in a new corner of our infrastructure and re-generated every time trimmability is tested. Put another way, I agree w/ the direction but suggest getting there by a slightly different route.

The targets in CodeGen.proj and the supporting bits at the bottom of ResolveReferences.targets could easily be extended to collect `$(IsTrimmable)` information. I leave it up to you whether you want to generate a brand new file in eng/ or add the information to one of eng/ProjectReferences.props or eng/SharedFramework.Local.props.

More generally, I'd like to be fairly sure we do _not_ limit checking assemblies are trimmable to the shared framework projects. I could see us wanting to ensure something that ships only in a package is and remains trimmable.

JamesNK committed:
Move linker app to tools and add some more projects

JamesNK commented:
@pranavkm I've moved the app to tools. Feel free to apply your changes.

I added a couple of projects, up until I hit one that generated trim warnings: `Microsoft.AspNetCore.Http.Abstractions`. I fixed some warnings, most are still around. I want to double-check the warnings show up in build.

JamesNK commented on src/Shared/PropertyHelper/PropertyHelper.cs:
```diff
@@ -136,7 +137,8 @@ public void SetValue(object instance, object? value)
     /// <returns>A cached array of all public properties of the specified type.
     /// </returns>
-    public static PropertyHelper[] GetProperties(TypeInfo typeInfo)
+    public static PropertyHelper[] GetProperties(
```
I think PropertyHelper is general might be incompatible with trimming.

JamesNK commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -95,6 +95,7 @@ public RouteValueDictionary()
     /// Only public instance non-index properties are considered.
     /// </remarks>
+    [RequiresUnreferencedCode("RouteValueDictionary uses reflection to generate the route values.")]
     public RouteValueDictionary(object? values)
```
I think we might want new public API ctors for `RouteValueDictionary`.

Copying values from a `IDictionary` is fine.  Copying values via reflection from public properties seems incompatible with trimming.

pranavkm commented:
@JamesNK lets do this in two passes - we can get this change in once we resolve the current failure and I'll send a PR to use @dougbu's suggestion to find projects that are interesting to test.

pranavkm commented on src/Tools/LinkabilityChecker/LinkabilityChecker.csproj:
```diff
@@ -0,0 +1,80 @@
+    <Reference Include="Microsoft.Extensions.Features" />
+    <Reference Include="Microsoft.Extensions.ObjectPool" />
+    <Reference Include="Microsoft.AspNetCore.Http.Features" />
+    <Reference Include="Microsoft.AspNetCore.Http.Abstractions" />
```
```suggestion
```

We can resolve this in a follow-up.

pranavkm committed:
Update src/Tools/LinkabilityChecker/LinkabilityChecker.csproj

pranavkm committed:
Merge branch 'main' into jamesnk/linkertest

JamesNK commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -95,6 +95,7 @@ public RouteValueDictionary()
     /// property names are keys, and property values are the values, and copied into the dictionary.
     /// Only public instance non-index properties are considered.
     /// </remarks>
+    [RequiresUnreferencedCode("RouteValueDictionary uses reflection to generate the route values.")]
```
```suggestion
```
Will look at this change again in followup PR to resolve Http.Abstractions warnings

JamesNK commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -744,6 +745,7 @@ internal class PropertyStorage
         public readonly object Value;
         public readonly PropertyHelper[] Properties;
 
+        [RequiresUnreferencedCode("PropertyStorage uses reflection to get property values.")]
```
```suggestion
```
Will look at this change again in followup PR to resolve Http.Abstractions warnings

JamesNK committed:
Apply suggestions from code review

BrennanConroy commented on src/Extensions/Features/src/Microsoft.Extensions.Features.csproj:
```diff
@@ -12,7 +12,7 @@ Microsoft.AspNetCore.Http.Features.FeatureCollection
     <IsAspNetCoreApp>true</IsAspNetCoreApp>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <PackageTags>aspnetcore</PackageTags>
-    <Nullable>enable</Nullable>
```
Why is nullable being removed?

JamesNK replied:
Nullable is now turned on by default for all projects. Removing because it's redundent.

JamesNK commented on src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.csproj:
```diff
@@ -13,7 +13,7 @@ Microsoft.AspNetCore.Http.HttpResponse</Description>
     <PackageTags>aspnetcore</PackageTags>
     <IsPackable>false</IsPackable>
-    <Nullable>enable</Nullable>
+    <Trimmable>true</Trimmable>
```
```suggestion
```

JamesNK committed:
Update src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.csproj
